packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -762,6 +776,289 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
+          },
+        });
+      }
+    });
+
+    const bulkDeleteResponse = bulkDeleteParams.length
+      ? await this.client.bulk({
+          refresh,
+          body: bulkDeleteParams,
+          require_alias: true,
+        })
+      : undefined;
+
+    // extracted to ensure consistency in the error results returned
+    let errorResult: BulkDeleteItemErrorResult;
+    const objectsToDeleteAliasesFor: ObjectToDeleteAliasesFor[] = [];
+
+    const savedObjects = expectedBulkDeleteMultiNamespaceDocsResults.map((expectedResult) => {
+      if (isLeft(expectedResult)) {
+        return { ...expectedResult.value, success: false };
+      }
+      const {
+        type,
+        id,
+        namespaces,
+        esRequestIndex: esBulkDeleteRequestIndex,
+      } = expectedResult.value;
+      // we assume this wouldn't happen but is needed to ensure type consistency
+      if (bulkDeleteResponse === undefined) {
+        throw new Error(
+          `Unexpected error in bulkDelete saved objects: bulkDeleteResponse is undefined`
+        );
+      }
+      const rawResponse = Object.values(
+        bulkDeleteResponse.items[esBulkDeleteRequestIndex]
+      )[0] as NewBulkItemResponse;
+
+      const error = getBulkOperationError(type, id, rawResponse);
+      if (error) {
+        errorResult = { success: false, type, id, error };
+        return errorResult;
+      }
+      if (rawResponse.result === 'not_found') {
+        errorResult = {
+          success: false,
+          type,
+          id,
+          error: errorContent(SavedObjectsErrorHelpers.createGenericNotFoundError(type, id)),
+        };
+        return errorResult;
+      }
+
+      if (rawResponse.result === 'deleted') {
+        // `namespaces` should only exist in the expectedResult.value if the type is multi-namespace.
+        if (namespaces) {
+          objectsToDeleteAliasesFor.push({
+            type,
+            id,
+            ...(namespaces.includes(ALL_NAMESPACES_STRING)
+              ? { namespaces: [], deleteBehavior: 'exclusive' }
+              : { namespaces, deleteBehavior: 'inclusive' }),
+          });
+        }
+      }
+      const successfulResult = {
+        success: true,
+        id,
+        type,
+      };
+      return successfulResult;
+    });
+
+    // Delete aliases if necessary, ensuring we don't have too many concurrent operations running.
+    const mapper = async ({ type, id, namespaces, deleteBehavior }: ObjectToDeleteAliasesFor) =>
+      await deleteLegacyUrlAliases({
+        mappings: this._mappings,
+        registry: this._registry,
+        client: this.client,
+        getIndexForType: this.getIndexForType.bind(this),
+        type,
+        id,
+        namespaces,
+        deleteBehavior,
+      }).catch((err) => {
+        this._logger.error(`Unable to delete aliases when deleting an object: ${err.message}`);
+      });
+    await pMap(objectsToDeleteAliasesFor, mapper, { concurrency: MAX_CONCURRENT_ALIAS_DELETIONS });
```

Comment:
As mentioned in https://github.com/elastic/kibana/pull/117056#discussion_r741844240, there isn't a bulk API for deleting legacy URL aliases for many objects since the es query/script would need to handle different namespaces and delete behavior for each saved object that no longer exists. To minimize the number of changes adding a `bulkDelete` API introduces, I kept it simple, I'm using a `pMap`, as is done in [`updateObjectSpaces`](https://github.com/elastic/kibana/blob/main/packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/update_objects_spaces.ts#L280).